crates/red_knot_python_semantic/src/types/infer.rs

```diff
@@ -2204,6 +2219,32 @@ impl<'db> TypeInferenceBuilder<'db> {
             DeferredExpressionState::from(self.are_all_types_deferred()),
         );
 
+        if target
+            .as_name_expr()
+            .is_some_and(|name| &name.id == "TYPE_CHECKING")
+        {
+            if !self.in_stub()
+                && !matches!(
+                    value
+                        .as_ref()
+                        .and_then(|value| value.as_boolean_literal_expr()),
+                    Some(ast::ExprBooleanLiteral { value: false, .. })
+                )
+            {
+                report_invalid_type_checking_constant(&self.context, assignment.into());
```

Comment:
I think we should report this diagnostic also in a stub file, if we see e.g. `TYPE_CHECKING: object = "str"` (and add a test for this.) So I think we should check the RHS (if any) regardless of whether we are in a stub or not, but we should also allow a `...` RHS in a stub.